### ISSUE TITLE
Do not suggest `unwrap` on `Some(T)` where `T` is expected

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#42746]
Currently 

```rust
let i: i32 = Some(42);
```

reports

```
2 |     let i: i32 = Some(42);
  |                  ^^^^^^^^ expected i32, found enum `std::option::Option`
  |
  = note: expected type `i32`
             found type `std::option::Option<{integer}>`
  = help: here are some functions which might fulfill your needs:
          - .unwrap()
          - .unwrap_or_default()
```

I think `Option` and `Result` are special enough to warrant customized error messages. In this case simply suggesting to remove the `Some`

**Comment by nikomatsakis** [comment]
See #42929 -- I think that we should not be giving such broad suggestions based solely on types. It's too hard to tell the difference of what is appropriate or inappropriate.

**Comment by zackmdavis** [comment]
It's worth stressing that while this makes sense for the `Some` _literal_ in the given example, in more complicated situations, it's entirely plausible for `.unwrap()` to be the right suggestion when a variable of type `Option<T>` is given where `T` is expected. (Discussion on #42929 seemed to blur the distinction.)

**Comment by estebank** [comment]
[Whitelist](https://play.rust-lang.org/?gist=bde08f8d87ee7bbcaf1fdd7699467a3b&version=nightly) implemented in #46461:

```
error[E0308]: mismatched types
 --> src/main.rs:2:14
  |
2 | let i: i32 = Some(42);
  |              ^^^^^^^^ expected i32, found enum `std::option::Option`
  |
  = note: expected type `i32`
             found type `std::option::Option<{integer}>`

error: aborting due to previous error
```
